lib/tree.rb

```diff
@@ -16,51 +16,225 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n)
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # def add(key, value)
+  #   new_node = TreeNode.new(key, value)
+  #   current = @root
+  #   if current.nil?
+  #     current = new_node
+  #   else
+  #     done = false
+  #     until done
+  #       if key <= current.key
+  #         if !current.left.nil?
+  #           current = current.left
+  #         else
+  #           current.left = new_node
+  #           done = true
+  #         end
+  #       else
+  #         if !current.right.nil?
+  #           current = current.right
+  #         else
+  #           current.right = new_node
+  #           done = true
+  #         end
+  #       end
+  #     end
+  #   end
+  #   return current
+  # end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n)
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # def find(key)
+  #   current_node = @root
+  #   if current_node.nil?
+  #     return nil
+  #   else
+  #     while true
+  #       if current_node.key == key
+  #         return current_node.value
+  #       elsif current_node.key > key
+  #         if current_node.left.nil?
+  #           return false
+  #         else
+  #           current_node = current_node.left
+  #         end
+  #       else
+  #         if current_node.right.nil?
+  #           return false
+  #         else
+  #           current_node = current_node.right
+  #         end
+  #       end
+  #     end
+  #   end
+  # end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(2n) or O(n)
   def height
```

Comment:
👍 , but again the space complexity is O(log n) for a balanced tree and O(n) for an unbalanced one.